contracts/BalanceTrackerBase.sol

```diff
@@ -196,95 +204,129 @@ abstract contract BalanceTrackerBase {
         // Get account balance
         uint256 balance = mapRequesterBalances[requester] + initAmount;
 
+        // Total requester delivery rate is number of requests coming to a selected mech
+        uint256 totalDeliveryRate = deliveryRate * numRequests;
+
         // Adjust account balance
-        balance = _adjustInitialBalance(requester, balance, maxDeliveryRate, paymentData);
+        balance = _adjustInitialBalance(requester, balance, totalDeliveryRate, paymentData);
         mapRequesterBalances[requester] = balance;
 
-        emit RequesterBalanceAdjusted(requester, maxDeliveryRate, balance);
-
-        _locked = 1;
+        emit RequesterBalanceAdjusted(requester, totalDeliveryRate, balance);
     }
 
     /// @dev Finalizes mech delivery rate based on requested and actual ones.
     /// @param mech Delivery mech address.
-    /// @param requester Requester address.
-    /// @param requestId Request Id.
-    /// @param maxDeliveryRate Requested max delivery rate.
-    function finalizeDeliveryRate(address mech, address requester, uint256 requestId, uint256 maxDeliveryRate) external {
+    /// @param requesters Requester addresses.
+    /// @param deliveredRequests Set of mech request Id statuses: delivered / undelivered.
+    /// @param mechDeliveryRates Corresponding set of actual charged delivery rates for each request.
+    /// @param requesterDeliveryRates Corresponding set of requester agreed delivery rates for each request.
+    function finalizeDeliveryRates(
+        address mech,
+        address[] memory requesters,
+        bool[] memory deliveredRequests,
+        uint256[] memory mechDeliveryRates,
+        uint256[] memory requesterDeliveryRates
+    ) external virtual {
         // Reentrancy guard
-        if (_locked > 1) {
+        if (locked) {
             revert ReentrancyGuard();
         }
-        _locked = 2;
+        locked = true;
 
         // Check for marketplace access
         if (msg.sender != mechMarketplace) {
             revert MarketplaceOnly(msg.sender, mechMarketplace);
         }
 
-        // Get actual delivery rate
-        uint256 actualDeliveryRate = IMech(mech).getFinalizedDeliveryRate(requestId);
+        uint256 numRequests = deliveredRequests.length;
```

Comment:
Process only requests that were delivered. This bool array is created in the mech marketplace to reflect which request Ids were actually delivered by the mech. This allows to process bulks and not get blocked by requests that were already delivered.